NanoProd/plugins/TauSpinnerTableProducer.cc

```diff
@@ -0,0 +1,246 @@
+        name.replace(pos, 1, "p");
+      pos = name.find("-");
+      if (pos != std::string::npos)
+        name.replace(pos, 1, "minus");
+      out.push_back(std::make_pair(name, val));
+    }
+    return out;
+  }
+
+  void printModuleInfo(edm::ParameterSet const &config) const {
+    std::string header = (boost::format("%-39s%39s") % config.getParameter<std::string>("@module_type") %
+                          config.getParameter<std::string>("@module_label"))
+                             .str();
+    std::cout << std::string(78, '-') << "\n";
+    std::cout << header << "\n";
+    // std::cout << "Produces: " << flow << "\n";
+    std::cout << boost::format("%-15s : %-60s\n") % "Input" % config.getParameter<edm::InputTag>("input").encode();
+    std::cout << boost::format("%-15s : %-60s\n") % "Branch" % config.getParameter<std::string>("branch");
+    std::string thetaStr;
+    for (const auto &theta : theta_vec_)
+      thetaStr += theta.first + ",";
+    std::cout << boost::format("%-15s : %-60s\n") % "Theta" % thetaStr;
+  }
+
+  edm::EDGetTokenT<edm::View<reco::GenParticle>> genPartsToken_;
+  std::string branch_;
+  std::vector<std::pair<std::string, double>> theta_vec_;
+  int bosonPdgId_;
+  std::string tauSpinnerPDF_;
+  bool ipp_;
+  int ipol_;
+  int nonSM2_;
+  int nonSMN_;
+  double cmsE_;
+};
+
+TauSpinnerTableProducer::TauSpinnerTableProducer(const edm::ParameterSet &config)
+    : genPartsToken_(consumes(config.getParameter<edm::InputTag>("input"))),
+      branch_(config.getParameter<std::string>("branch")),
+      bosonPdgId_(25),                                             //Higgs, to be configurable?
+      tauSpinnerPDF_(config.getParameter<std::string>("pdfSet")),  //"NNPDF30_nlo_as_0118"
+      ipp_(true),                                                  //pp collisions
+      ipol_(0),
+      nonSM2_(0),
+      nonSMN_(0),
+      cmsE_(config.getParameter<double>("cmsE"))  //cms energy in GeV, 13000.0
+{
+  theta_vec_ = nameAndValue(config.getParameter<std::vector<double>>("theta"));
+
+  printModuleInfo(config);
+
+  //state that we use tauola/tauspinner resource
+  usesResource(edm::SharedResourceNames::kTauola);
+
+  produces<nanoaod::FlatTable>();
+}
+
+reco::GenParticle TauSpinnerTableProducer::getBoson(const edm::View<reco::GenParticle> &parts, bool &foundBoson) const {
+  reco::GenParticle boson;
+  foundBoson = false;
+  for (auto part : parts) {
+    if (std::abs(part.pdgId()) == bosonPdgId_ && part.isLastCopy()) {
+      boson = part;
+      foundBoson = true;
+      break;
+    }
+  }
+  return boson;
+}
+
+reco::GenParticleRef TauSpinnerTableProducer::getLastCopy(const reco::GenParticleRef &part) const {
+  reco::GenParticleRef last_copy(part);
+  if (part->statusFlags().isLastCopy())
+    return last_copy;
+  for (const auto &daughter : part->daughterRefVector()) {
+    if (daughter->pdgId() == part->pdgId()) {
+      last_copy = getLastCopy(daughter);
+    }
+  }
+  return last_copy;
```

Comment:
this will not work correctly for all events. You need to add requirements on status flags. For example:
```
+-> tau+ <-15> pt=69.1271 eta=1.86289 phi=1.96457 E=228.042 status=23 statusFlags=101100110000001
     +-> tau+ <-15> pt=16.8648 eta=1.55281 phi=1.33991 E=41.6633 status=2 statusFlags=010000100000011
     |   +-> nu_taubar <-16> pt=0.886862 eta=1.37225 phi=0.973728 E=1.86141 index=39 status=1 statusFlags=011011000111100
     |   +-> e+ <-11> pt=14.8502 eta=1.54707 phi=1.35188 E=36.4614 status=1 statusFlags=011011000111100
     |   +-> nu_e <12> pt=1.1958 eta=1.68676 phi=1.45717 E=3.34049 status=1 statusFlags=011011000111100
     +-> gamma <22> pt=3.22423 eta=1.89204 phi=1.93187 E=10.936 status=1 statusFlags=011000000000001
     +-> tau- <15> pt=48.0004 eta=2.04085 phi=2.1706 E=187.86 status=2 statusFlags=010000000000011
     |   +-> nu_tau <16> pt=13.0619 eta=1.98695 phi=2.15928 E=48.5272 status=1 statusFlags=011000000111100
     |   +-> pi0 <111> pt=31.8674 eta=2.05696 phi=2.17246 E=126.672 status=2 statusFlags=111000000111110
     |   |   +-> gamma <22> pt=26.9922 eta=2.05856 phi=2.17325 E=107.461 status=1 statusFlags=011000001001100
     |   |   +-> gamma <22> pt=4.87527 eta=2.04799 phi=2.16808 E=19.2118 status=1 statusFlags=011000001001100
     |   +-> pi- <-211> pt=3.07326 eta=2.09377 phi=2.1995 E=12.6606 status=1 statusFlags=011000000111100
     +-> gamma <22> pt=0.00031434 eta=2.25028 phi=2.26373 E=0.00150816 status=1 statusFlags=011000000000001
     +-> tau+ <-15> pt=2.95819 eta=2.05438 phi=1.84956 E=11.8633 statusFlags=010000000000011
         +-> nu_taubar <-16> pt=1.8583 eta=1.95522 phi=1.47631 E=6.69638 status=1 statusFlags=011000000111100
         +-> pi0 <111> pt=0.332265 eta=2.32482 phi=1.59947 E=1.72023 status=2 statusFlags=111000000111110
         |   +-> gamma <22> pt=0.288605 eta=2.1868 phi=1.52188 E=1.30145 status=1 statusFlags=011000001001100
         |   +-> e- <11> pt=0.0351559 eta=2.92245 phi=1.57597 E=0.327664 status=1 statusFlags=011000001001100
         |   +-> e+ <-11> pt=0.025022 eta=1.96615 phi=2.78592 E=0.0911206 status=1 statusFlags=011000001001100
         +-> pi+ <211> pt=1.18237 eta=1.73136 phi=2.54749 E=3.44671 status=1 statusFlags=011000000111100
```

Reply:
it is a "normal" tau decay - of course not very frequent, but still present in H->tautau samples. I agree that the Pythia decay tree is a bit confusing, but it is what it is, and we need to handle it...

What is happening here is tau emits an FSR photon that splits into two leptons. Phytia omits this virtual split photon and attaches the split l+l- directly to the tau mother. It is more frequent that photon splits into e+e- or mu+mu-, but as in the example above, sometimes it is tau+tau-. The "real" tau daughter can be identified by the `kFromHardProcess` status flag (bit 8). As you can see only one out of 3 tau daughters have this flag set.

By the way, the [GenLepton](https://github.com/cms-tau-pog/TauMLTools/blob/master/Analysis/interface/GenLepton.h) class from TauMLTools correctly handles these cases. At some point, we can consider integrating it into CMSSW, because, as we discussed in the past, the current gen tau DM implementation in CMSSW also doesn't handle all corner cases correctly.

Reply:
Thanks for explanation, now I see the point.
What concerns including handling such cases in official CMSSW: it will be good to made some effort to have it rather soon to be sure it is implemented for next NanoAOD - it should be feasible if problem is already recognized and  (at least partly) implemented in TauMLTools. But it is beyond this PR.